lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/TitleAndButtonsContainer.kt

```diff
@@ -0,0 +1,181 @@
+
+    fun setBackgroundColor(color: Colour) = if (color.hasValue()) setBackgroundColor(color.get()) else Unit
+
+    fun setTitleFontSize(size: Float) = this.titleSubTitleBar.setTitleFontSize(size)
+
+    fun setTitleTypeface(typefaceLoader: TypefaceLoader, font: FontOptions) = this.titleSubTitleBar.setTitleTypeface(typefaceLoader, font)
+
+    fun setSubtitleTypeface(typefaceLoader: TypefaceLoader, font: FontOptions) = this.titleSubTitleBar.setSubtitleTypeface(typefaceLoader, font)
+
+    fun setSubtitleFontSize(size: Float) = this.titleSubTitleBar.setSubtitleFontSize(size)
+
+    fun setSubtitleColor(@ColorInt color: Int) = this.titleSubTitleBar.setSubtitleTextColor(color)
+
+    fun setTitleColor(@ColorInt color: Int) = this.titleSubTitleBar.setTitleTextColor(color)
+
+    fun getTitle(): String = this.titleSubTitleBar.getTitle()
+
+    fun clearCurrentTitle() {
+        this.titleSubTitleBar.clear()
+        clearComponent()
+    }
+
+    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
+        val titleComponent = getTitleComponent()
+        val isCenter = titleComponentAlignment == Alignment.Center
+        val containerWidth = r - l
+        val containerHeight = b - t
+        val isRTL = isRTL()
+        val titleWidth = titleComponent.measuredWidth
+        val titleHeight = titleComponent.measuredHeight
+        val leftBarWidth = leftButtonBar.measuredWidth
+        val rightBarWidth = rightButtonBar.measuredWidth
+
+        val (titleLeft, titleRight) = resolveHorizontalTitleBoundsLimit(containerWidth, titleWidth, leftBarWidth, rightBarWidth, isCenter, isRTL)
+        val (titleTop, titleBottom) = resolveVerticalTitleBoundsLimit(containerHeight, titleHeight)
+        val (leftButtonsLeft, leftButtonsRight) = resolveLeftButtonsBounds(containerWidth, leftBarWidth, isRTL)
+        val (rightButtonsLeft, rightButtonsRight) = resolveRightButtonsBounds(containerWidth, rightButtonBar.measuredWidth, isRTL)
+
+        leftButtonBar.layout(leftButtonsLeft, t, leftButtonsRight, b)
+        rightButtonBar.layout(rightButtonsLeft, t, rightButtonsRight, b)
+        titleComponent.layout(titleLeft, titleTop, titleRight, titleBottom)
+
+        measureTextTitleEnsureTruncatedEndIfNeeded(titleRight, titleLeft, titleComponent, isCenter)
+    }
+
+    private fun measureTextTitleEnsureTruncatedEndIfNeeded(titleRight: TitleRight, titleLeft: TitleLeft, titleComponent: View, isCenter: Boolean) {
+        if (component == null && titleRight - titleLeft != titleComponent.measuredWidth) {
+            val margin = if (isCenter) 0 else 2 * DEFAULT_LEFT_MARGIN_PX
+            titleComponent.measure(MeasureSpec.makeMeasureSpec(titleRight - titleLeft + margin, MeasureSpec.EXACTLY),
+                    MeasureSpec.makeMeasureSpec(titleComponent.measuredHeight, MeasureSpec.EXACTLY))
+        }
+    }
+
+    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
```

Comment:
Extracting all of the measurements and layout concerns to dedicated classes would have allowed you to write cleaner and shorter tests.

Reply:
The tests are short mainly not that long, and the logic for measurements was extracted to a separate high order function where its tests are really short.